UCX fails to compile with GCC-15

### Describe the bug
GCC-15 defaults to C23. It made illegal some constructs that are used in ucx are made illegal, for example K&R-style empty function parameters are now interpreted as void and meant that there are no parameters, passing such functions to call sites result in `incompatible pointer types` compilation error.

### Steps to Reproduce
make -j8 --shuffle
- using ucx-1.13.1, but problem code still present in 1.18.0
- 
```
configure: =========================================================
configure: UCX build configuration:
configure:         Build prefix:   /usr
configure:    Configuration dir:   /etc/ucx
configure:   Preprocessor flags:   -DCPU_FLAGS="" -I${abs_top_srcdir}/src -I${abs_top_builddir} -I${abs_top_builddir}/src
configure:           C compiler:   x86_64-pc-linux-gnu-gcc -g -Wall -funwind-tables -Wno-missing-field-initializers -Wno-unused-parameter -Wno-unused-label -Wno-long-long -Wno-endif-labels -Wno-sign-compare -Wno-multichar -Wno-deprecated-declarations -Winvalid-pch -Wno-pointer-sign -Werror-implicit-function-declaration -Wno-format-zero-length -Wnested-externs -Wshadow -Werror=declaration-after-statement
configure:         C++ compiler:   x86_64-pc-linux-gnu-g++ -g -Wall -funwind-tables -Wno-missing-field-initializers -Wno-unused-parameter -Wno-unused-label -Wno-long-long -Wno-endif-labels -Wno-sign-compare -Wno-multichar -Wno-deprecated-declarations -Winvalid-pch
configure:         Multi-thread:   disabled
configure:         NUMA support:   enabled
configure:            MPI tests:   disabled
configure:          VFS support:   no
configure:        Devel headers:   no
configure: io_demo CUDA support:   no
configure:             Bindings:   < >
configure:          UCS modules:   < >
configure:          UCT modules:   < cma >
configure:         CUDA modules:   < >
configure:         ROCM modules:   < >
configure:           IB modules:   < >
configure:          UCM modules:   < >
configure:         Perf modules:   < >
configure: =========================================================
```

### Setup and versions
- Gentoo x86_64
   - Linux nho-laptop 6.14.0-gentoo #1 SMP PREEMPT Tue Mar 25 12:58:44 MSK 2025 x86_64 AMD Ryzen 7 5700U with Radeon Graphics AuthenticAMD GNU/Linux


### Additional information (depending on the issue)
See Gentoo bug https://bugs.gentoo.org/944992

Commit d58d2ae24b9e8498775a2ad0806f1421828fa935 fails to build with:

```
arch/x86_64/cpu.c:502:5: error: conflicting types for 'ucs_arch_get_cpu_flag' due to enum/integer mismatch; have 'int(void)' [-Werror=enum-int-mismatch]
  502 | int ucs_arch_get_cpu_flag()
      |     ^~~~~~~~~~~~~~~~~~~~~
In file included from /home/nho/workshop/ucx/src/ucs/arch/cpu.h:105,
                 from arch/x86_64/cpu.c:15:
/home/nho/workshop/ucx/src/ucs/arch/x86_64/cpu.h:51:16: note: previous declaration of 'ucs_arch_get_cpu_flag' with type 'ucs_cpu_flag_t(void)' {aka 'enum ucs_cpu_flag(void)'}
   51 | ucs_cpu_flag_t ucs_arch_get_cpu_flag() UCS_F_NOOPTIMIZE;
      |                ^~~~~~~~~~~~~~~~~~~~~
```

--- a/AUTHORS
+++ b/AUTHORS
@@ -57,6 +57,7 @@ Khaled Hamidouche <Khaled.Hamidouche@amd.com>
 Konstantin Belousov <kib@freebsd.org>
 Leonid Genkin <lgenkin@nvidia.com>
 Lior Paz <liorpa@nvidia.com>
+Lisandro Dalcin <dalcinl@gmail.com>
 Luis E. Pena <l31g@hotmail.com>
 Manjunath Gorentla Venkata <manjugv@gmail.com>
 Marek Schimara <Marek.Schimara@bull.net>

--- a/src/tools/perf/api/libperf.h
+++ b/src/tools/perf/api/libperf.h
@@ -12,12 +12,12 @@
 
 #include <ucs/sys/compiler.h>
 
+#include <uct/api/uct.h>
+#include <ucp/api/ucp.h>
+
 BEGIN_C_DECLS
 
 /** @file libperf.h */
-
-#include <uct/api/uct.h>
-#include <ucp/api/ucp.h>
 
 
 typedef enum {

--- a/src/tools/perf/lib/libperf_int.h
+++ b/src/tools/perf/lib/libperf_int.h
@@ -11,10 +11,6 @@
 
 #include <tools/perf/api/libperf.h>
 
-BEGIN_C_DECLS
-
-/** @file libperf_int.h */
-
 #include <ucs/async/async.h>
 #include <ucs/time/time.h>
 #include <ucs/sys/math.h>
@@ -23,6 +19,10 @@ BEGIN_C_DECLS
 #if _OPENMP
 #include <omp.h>
 #endif
+
+BEGIN_C_DECLS
+
+/** @file libperf_int.h */
 
 
 #define TIMING_QUEUE_SIZE    2048

--- a/src/ucs/arch/aarch64/cpu.h
+++ b/src/ucs/arch/aarch64/cpu.h
@@ -150,7 +150,7 @@ static inline ucs_cpu_model_t ucs_arch_get_cpu_model()
     return UCS_CPU_MODEL_ARM_AARCH64;
 }
 
-static inline int ucs_arch_get_cpu_flag()
+static inline ucs_cpu_flag_t ucs_arch_get_cpu_flag()
 {
     return UCS_CPU_FLAG_UNKNOWN;
 }

--- a/src/ucs/arch/ppc64/cpu.h
+++ b/src/ucs/arch/ppc64/cpu.h
@@ -61,7 +61,7 @@ static inline ucs_cpu_vendor_t ucs_arch_get_cpu_vendor()
     return UCS_CPU_VENDOR_GENERIC_PPC;
 }
 
-static inline int ucs_arch_get_cpu_flag()
+static inline ucs_cpu_flag_t ucs_arch_get_cpu_flag()
 {
     return UCS_CPU_FLAG_UNKNOWN;
 }

--- a/src/ucs/arch/rv64/cpu.h
+++ b/src/ucs/arch/rv64/cpu.h
@@ -62,7 +62,7 @@ static inline ucs_cpu_model_t ucs_arch_get_cpu_model()
     return UCS_CPU_MODEL_RV64G;
 }
 
-static inline int ucs_arch_get_cpu_flag()
+static inline ucs_cpu_flag_t ucs_arch_get_cpu_flag()
 {
     return UCS_CPU_FLAG_UNKNOWN;
 }

--- a/src/ucs/arch/x86_64/cpu.c
+++ b/src/ucs/arch/x86_64/cpu.c
@@ -499,9 +499,9 @@ ucs_cpu_model_t ucs_arch_get_cpu_model()
 }
 
 
-int ucs_arch_get_cpu_flag()
+ucs_cpu_flag_t ucs_arch_get_cpu_flag()
 {
-    static int cpu_flag = UCS_CPU_FLAG_UNKNOWN;
+    static ucs_cpu_flag_t cpu_flag = UCS_CPU_FLAG_UNKNOWN;
 
     if (UCS_CPU_FLAG_UNKNOWN == cpu_flag) {
         uint32_t result = 0;
@@ -553,7 +553,7 @@ int ucs_arch_get_cpu_flag()
                 result |= UCS_CPU_FLAG_AVX2;
             }
         }
-        cpu_flag = result;
+        cpu_flag = (ucs_cpu_flag_t) result;
     }
 
     return cpu_flag;

--- a/src/ucs/sys/init.c
+++ b/src/ucs/sys/init.c
@@ -32,7 +32,7 @@ static UCS_F_NOOPTIMIZE void ucs_check_cpu_flags(void)
 {
     char str[256];
     char *p_str;
-    int cpu_flags;
+    ucs_cpu_flag_t cpu_flags;
     struct {
         const char* flag;
         ucs_cpu_flag_t value;

